### PR TITLE
Added tracking for objects that are referenced between summaries.

### LINKED ISSUE #7924: Identify and update GC when objects are referenced
Work item for the problem and design described here - https://github.com/microsoft/FluidFramework/issues/7493.

We need to determine when an object is referenced so that we can reset the `unreferencedTimer`, if any, on the object and it's dependencies.
This is needed because GC runs on the state of the system at summary time. Today, it cannot determine changes that happens between 2 GC runs, such as an object transitioning from unreferenced -> referenced -> unreferenced. Without this knowledge, it will not reset the `unreferencedTime` for such objects and may end up deleting them prematurely leading to bugs.

### PR DESCRIPTION
**Pull request by agarwal-navin**
This will help identify objects that transition from `unreferenced -> referenced -> unreferenced` between summaries. The unreferenced timestamp of these objects needs to be updated since a client could have added in-memory references to them. We will need to adjust the time to wait before delete accordingly.

This change does the following:
- Every time a handle is parsed, the serializer notifies the root context (container runtime) of its URL.
- During GC, get objects that have been referenced since the last GC run. Then find objects that transitioned from `unreferenced -> referenced -> unreferenced` between the last and the current GC run.
- Reset the unreferenced timestamp of these objects.
- The new timestamp will be added when the GC mark phase runs.

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -1673,6 +1676,60 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
+     * that transitioned from `unreferenced -> referenced -> unreferenced` since the last GC run.
+     * @param resultFromCurrentRun - The result of the current GC run.
+     */
+    private resetUnreferencedState(resultFromCurrentRun: IGCResult) {
+        assert(this.gcDataFromLastRun !== undefined, "GC data from last run should be available");
+
+        // Get the list of references that were added since the last GC run.
+        // TODO: The cast to any is done temporarily until the change to IFluidHandleContext in `core-interfaces` is
+        // merged and client is updated to the new `core-interfaces` version.
+        const outboundRoutesSinceLastGC: string[] = (this.IFluidHandleContext as any).decodedHandles;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1690]: Note: The cast to any will be removed once the change in IFluidHandleContext is merged and the client is updated to use the new `core-interfaces` bits.

**Comment by agarwal-navin** [comment]
In this PR, the serializer updates the container runtime when a handle is parsed.

An alternate approach could be to save the decode handles in the serializer. Then expose an API that returns these handles. The container runtime can then get this information from each data store which gets it from its DDSs. We can either update the `getGCData` API to return this information or add a new API. This will keep the data localized to each DDS and make the implementation cleaner.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -1718,6 +1787,9 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             if (this.gcTestMode) {
                 this.dataStores.deleteUnusedRoutes(deletedNodeIds);
             }
+
+            // Update the GC data as per this run.
+            this.gcDataFromLastRun = cloneGCData(gcData);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1792]: This should not be updated here. This should only happen when we get the ack for the summary, similar to how we track other start across summaries.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 3]
```diff
@@ -1768,6 +1840,11 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         assert(summarizeResult.summary.type === SummaryType.Tree,
             0x12f /* "Container Runtime's summarize should always return a tree" */);
 
+        // Reset the decodeHandles. We need to keep track of handles decoded between two summaries.
+        // TODO: The cast to any is done temporarily until the change to IFluidHandleContext in `core-interfaces` is
+        // merged and client is updated to the new `core-interfaces` version.
+        (this.IFluidHandleContext as any).decodedHandles = [];
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1846]: This should not be updated here. This should only happen when we get the ack for the summary, similar to how we track other start across summaries.

File: packages/runtime/runtime-utils/src/serializer.ts [hunk 4]
```diff
@@ -101,6 +101,10 @@ export class FluidSerializer implements IFluidSerializer {
                 ? value.url
                 : generateHandleContextPath(value.url, this.context);
 
+            // Update the root context of all handles that are decoded.
+            // TODO: The cast to any is done temporarily until the change to IFluidHandleContext in `core-interfaces` is
+            // merged and client is updated to the new `core-interfaces` version.
+            (this.root as any).handleDecoded?.(absolutePath);
```
[the remark this answers is not in the document]
Comment (vladsud) [reply packages/runtime/runtime-utils/src/serializer.ts:107]: I think you are saying (and maybe I misunderstood comment initially) that you will break this PR into two, and it will be done properly, right? If so, it's better, but I think we need to still figure out next step - how to make sure this is not optional (in 2 releases), so comment and follow up issues should track it

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1690]: This does not look right, given that this.IFluidHandleContext is of type IFluidHandleContext which does not have decodedHandles. You rely here on concrete class implementation, and there is no guarantee that is what it is.

File: common/lib/core-interfaces/src/handles.ts [hunk 5]
```diff
@@ -39,6 +39,12 @@ export interface IFluidHandleContext extends IProvideFluidHandleContext {
     attachGraph(): void;
 
     resolveHandle(request: IRequest): Promise<IResponse>;
+
+    /**
+     * Called when a handle belonging to this context is decoded.
+     * @param url - The url of the decoded handle.
+     */
+    handleDecoded(url: string): void;
```
Comment (vladsud) [inline common/lib/core-interfaces/src/handles.ts:47]: This object for most part had nothing to do with GC till now. It was used for creation of paths and resolving URIs.
Is there a way to keep GC & routing as a separate concerns?
They overlap for sure (and former depends on latter), but can latter have only one responsibility and not get tied directly into GC interactions?

I believe what we deal with here is a contract - a requirement to IFluidDataStore to do something (track new paths) and expose it to GC at a later time. If so, should it be explicit requirement (API) on IFluidDataStoreChannel, and not this hidden and very implicit requirement to call this API?
I.e. if I'm a developer building implementation of IFluidDataStoreChannel, how would I know that I need to call it? when?

File: packages/runtime/container-runtime/src/containerHandleContext.ts [hunk 6]
```diff
@@ -42,4 +45,8 @@ export class ContainerFluidHandleContext implements IFluidHandleContext {
     public async resolveHandle(request: IRequest): Promise<IResponse> {
         return this.runtime.resolveHandle(request);
     }
+
+    public handleDecoded(url: string): void {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerHandleContext.ts:49]: It feels to me that it's better to track handles themselves than URIs.
Or if we do URIs, we will need to clearly prescribe what URI means.
For example, if https://github.com/microsoft/FluidFramework/issues/6001 were to happen, will it be full URI, or first part of it? What if handles evolve over time and are structured somehow differently (not sure how, but let's say they have 3 parts :)). I feel layers like that should not care, and only GC should reason about what info it needs to extract from handle.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 7]
```diff
@@ -1673,6 +1676,60 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         return this.context.getAbsoluteUrl(relativeUrl);
     }
 
+    /**
+     * Find and reset nodes whose unreferenced state, such as unreferenced time, needs to be reset. These are nodes
+     * that transitioned from `unreferenced -> referenced -> unreferenced` since the last GC run.
+     * @param resultFromCurrentRun - The result of the current GC run.
+     */
+    private resetUnreferencedState(resultFromCurrentRun: IGCResult) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1684]: Any chance we can group all GC-related methods, variables and refactor it into its own class, such that ContainerRuntime only need to know about few entry points, like collectGarbage()?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1690]: Oh, here it might be fine (I was looking at data store level), but you can avoid it by changing type of IFluidHandleContext() getter to return concrete class, or having internal variable of right type

File: common/lib/core-interfaces/src/handles.ts (hunk 5 above)
Comment (agarwal-navin) [reply common/lib/core-interfaces/src/handles.ts:47]: I do not like this approach as well. I proposed something along the lines of what are you saying here - https://github.com/microsoft/FluidFramework/pull/8053#issuecomment-954991925

I think we can extend the API surface of the IFluidSerializer by adding two more - getEncodedHandles() and getDecodedHandles() which would returns all handles serialized and parsed during its lifetime. We can then use this information accordingly for GC.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1684]: I was actually doing something similar but in summarizer node - Add a RootSummarizerNode class which handles all GC related functionality of the container runtime. That is turning out to be more complicated than what I want. So maybe creating a class off of container runtime might be simpler.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 8]
```diff
@@ -1673,6 +1676,60 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
+        // TODO: The cast to any is done temporarily until the change to IFluidHandleContext in `core-interfaces` is
+        // merged and client is updated to the new `core-interfaces` version.
+        const outboundRoutesSinceLastGC: string[] = (this.IFluidHandleContext as any).decodedHandles;
+
+        /**
+         * Run GC on the data from the last run to get the reference graph starting from the above routes.
+         * Note that this has to be done on the reference graph from the last run because routes added since then may
+         * have had outbound references that have been removed now. We need to get a list of all those references.
+         * For example, consider the following scenario where A, B and C are data stores:
+         * 1. Last summary happened at t1. Reference graph: A    B (t1) -> C (t1). B and C have unreferenced time t1.
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1697]: It was not clear for me what you mean here initially.
The reason - you are listing nodes and edges in same list, which is a bit confusing (even comma after A would help).
I'd rather see something more closer to how it's represented in math (see [example](https://en.wikipedia.org/wiki/Graph_(discrete_mathematics)#Examples)):
V = [A, B, C], E = [{B, C}] (they are sets, so in math we use {}, not [] :))
For readability we can describe it as E = [B -> C], but E has no A as there is no edges from A.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 9]
```diff
@@ -1685,13 +1742,25 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
                 deletedDataStores?: number,
                 totalDataStores?: number,
             } = {};
+
+            // If we are running GC for the first time in this session, get the GC data from the last run from the
+            // initial GC data in base summary.
+            if (this.gcDataFromLastRun === undefined) {
+                this.gcDataFromLastRun = await this.dataStores.getInitialGCData();
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1749]: Can this be done in constructor, to make gcDataFromLastRun non-optional?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1792]: So, what's next? :) I.e. are you planning to do this in this PR, or open an issue for future, or something else?

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1792]: I should have probably mentioned that :-). This is a pain every time we have something new to track. I am looking at how to simplify this. Maybe have a separate class like `RootSummarizerNode` that tracks all of this stuff for container runtime. Or maybe we can simplify this tracking process. But yeah, that has to happen before this PR can go in.
So, I will update this either in this PR or create a separate PR for this.

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 10]
```diff
@@ -458,6 +461,28 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         return this.channel.getGCData(fullGC);
     }
 
+    /**
+     * After GC has run, called to notify this data stores of routes whose unreferenced state needs to be reset. These
+     * are routes to nodes in the data store that have transitioned from `unreferenced -> referenced -> unreferenced`
+     * since the last GC run.
+     * @param routesToReset - The routes to reset in this data store.
+     */
+    public resetUnreferencedState(routesToReset: string[]) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:470]: I wonder if it should be responsibility of "central GC mechanism" to track all the state in summaries (like last referenced state) vs. push this responsibility to data stores (if that is possible, of course).
I.e. ideally data it's not hard to implement data store contract. And if each data store needs to do the same thing, then we either need to consider moving it out and parent later (container runtime doing it), or provide stock implementation as a building block (hard in this case due to nesting).
Basically - instead of data stores having .gc nodes in summaries, can container runtime summarize and manage all that data in parallel summary tree?

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 11]
```diff
@@ -475,41 +500,66 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
...
+     * 1. When the unreferenced state of the data store is reset and the data store is loaded.
+     * 2. When the data store is realized and there was a GC run before that.
+     */
+    private resetChannelUnreferencedState(routes: string[]) {
+        assert(this.loaded, "Channel should be loaded when resetting unreferenced state");
+        assert(this.channel !== undefined, "Channel should be present when data store is loaded");
 
         // Remove the route to this data store, if it exists.
-        const usedChannelRoutes = this.lastUsedState.usedRoutes.filter(
+        const channelResetRoutes = routes.filter(
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:558]: Can this filtering happen in caller (resetUnreferencedState) such that there is no need to have that function - both callers could call this.channel.resetUnreferencedState directly

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 12]
```diff
@@ -460,10 +460,49 @@ export class DataStores implements IDisposable {
...
     }
 
+    /**
+     * Gets the GC data from the initial (base) summary this container loaded from.
+     */
+    public async getInitialGCData(): Promise<IGarbageCollectionData> {
+        const builder = new GCDataBuilder();
+        // Iterate over each store and get their initial GC data.
+        await Promise.all(Array.from(this.contexts)
+            .filter(([_, context]) => {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStores.ts:474]: .filter(([_, context]) => context.attachState === AttachState.Attached)

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 13]
```diff
@@ -460,10 +460,49 @@ export class DataStores implements IDisposable {
...
+     */
+    public resetUnreferencedState(routesToReset: string[]) {
+        // Get a map of data store ids to routes in it.
+        const dataStoreRoutesToReset = getChildNodesUsedRoutes(routesToReset);
+
+        // Call individual data stores to reset their unreferenced state.
+        for (const [id, routes] of dataStoreRoutesToReset) {
+            const dataStore = this.contexts.get(id);
+            assert(dataStore !== undefined, "Route does not belong to any known data store");
+            dataStore.resetUnreferencedState(routes);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStores.ts:502]: same question as above - how much of that logic can be moved to other layers (container runtime, summary nodes) such that implementation of alternative data store is possible (not a monster project that nobody other than us can do it)

File: packages/runtime/runtime-utils/src/serializer.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/serializer.ts:107]: Yep, I will break this PR into two once these changes look good.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1690]: I will update it.

File: packages/runtime/container-runtime/src/containerHandleContext.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerHandleContext.ts:49]: That makes sense. I can change this to handles.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 14]
```diff
@@ -123,8 +123,7 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
     }
 
     public get clientDetails(): IClientDetails {
-        // back-compat 0.38 - clientDetails is added to IFluidDataStoreContext in 0.38.
-        return this.dataStoreContext.clientDetails ?? this.dataStoreContext.containerRuntime.clientDetails;
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:127]: Can we remove IContainerRuntime(Base).clientDetails ? DataStoreContext works with concrete ContainerRuntime class, so it does not care.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 15]
```diff
@@ -140,10 +139,11 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
     }
 
     public get routeContext(): IFluidHandleContext {
-        // back-compat 0.38 - IFluidHandleContext is added to IFluidDataStoreContext in 0.38.
-        return this.dataStoreContext.IFluidHandleContext ?? this.dataStoreContext.containerRuntime.IFluidHandleContext;
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:144]: Same question here as above

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 16]
```diff
@@ -191,8 +191,7 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
         super();
 
         this.logger = ChildLogger.create(
-            // back-compat 0.38 - logger is added to IFluidDataStoreContext in 0.38.
-            dataStoreContext.logger ?? dataStoreContext.containerRuntime.logger,
+            dataStoreContext.logger,
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:194]: Same?

File: packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts [hunk 17]
```diff
@@ -460,18 +474,18 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
...
         // If GC is disabled, we are not tracking used state, return false.
         if (this.gcDisabled) {
             return false;
         }
 
-        return this.referenceUsedRoutes === undefined ||
-            JSON.stringify(this.usedRoutes) !== JSON.stringify(this.referenceUsedRoutes);
+        return this.referenceUsedRoutes === undefined
+            || JSON.stringify(this.usedRoutes) !== JSON.stringify(this.referenceUsedRoutes)
+            || this.stateResetSinceLastRun;
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts:488]: It's easier to follow code if terms used to describe things are the same.
It feels like hasUsedStateChanged() and stateResetSinceLastRun trying to answer same question, no?

**Comment by vladsud** [comment]
I might have missed it, but do we validate correctness by ensuring that new handles are actually noticed in ops?
I.e. that new route can't appear in summary # 2 if it was not present in summary # 1 and was not reported in ops in between these two summaries? I think it will be crucial to enforcing correctness of the system (and ideally test this mechanism in UTs)

**Comment by vladsud** [comment]
> In this PR, the serializer updates the container runtime when a handle is parsed.
> 
> An alternate approach could be to save the decode handles in the serializer. Then expose an API that returns these handles. The container runtime can then get this information from each data store which gets it from its DDSs. We can either update the `getGCData` API to return this information or add a new API. This will keep the data localized to each DDS and make the implementation / contracts cleaner.
> 
> Thoughts? @vladsud

I think the most important thing (no matter how we go about it) to clearly express expectations to data store developers on what is expected from them, when they implement interfaces (instead of using stock implementations). And in general, the more explicit the interaction ("you have to implement API X" vs. "you need to call a method on interface passed to you, if you do not, things will fall apart 30 days later") is better.

That said, I believe PR does save handles in serialize (indirectly, through handle context), no?

File: common/lib/core-interfaces/src/handles.ts (hunk 5 above)
Comment (vladsud) [reply common/lib/core-interfaces/src/handles.ts]: I think serializer can continue to be single purpose object, but it could take a callback on creation to notify someone about handles (or have eventing mechanism for that). Yes, this seems to be a bit better, but please take a look at comments about being more explicit - how do we ensure that developer who decides to reimplement most of these interfaces knows what he/she needs to do? I.e. how does system either forces developer (through compilation errors, assuming usage of TS) to do right things, or we have very clear documentation and protection in the code that very quickly tells developers they are missing something?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 3 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: I guess same question as above (on next steps to make your comment not be needed)

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1688]: Created an issue for this - https://github.com/microsoft/FluidFramework/issues/8126. I will be working on this before I merge this PR.

File: common/lib/core-interfaces/src/handles.ts (hunk 5 above)
Comment (agarwal-navin) [reply common/lib/core-interfaces/src/handles.ts]: I am not sure how the class taking a callback solves this problem. We can add documentation and comments but it does not define any contract right. Basically, since it's not built into the IFluidSerializer interface, it does not force developers to implement this functionality. We will need to rely on documentation and example implementations to do this.
